### DISCUSSION TITLE
Return generated id of fresh new aggregate.

### DISCUSSION THREAD

**Discussion by grzybsonssg** [Eventuous/eventuous#329] [Q&A]
I have a case when I want to generate ID of aggregate in my service code. But then, somehow I need to return it to client, so it can further interact with that aggregate. Here is a simple example of what I need:

``` c#
public class CardService
  : CommandService<Card, CardState, CardId> {
    public BookingService(IAggregateStore store) : base(store) {
        OnNew<Commands.CreateNewCard>(
            cmd => CardIdGenerator.GetNew(),
            (card, cmd)
                => card.Create(expiration: DateTimeOffset.UtcNow.AddYears(3))
        );
        
        OnExist<Commands.ChargeCard>(
            cmd => cmd.CardId,
            (card, cmd)
                => card.Charge(amount: cmd.Amount)
        );
    }
}
// ------
CommandService<Card, CardState, CardId> service;
await service.Handle(new CreateNewCard);
```

In such configuration, when I call `Handle` on command service, it will return `CardState`. Since 
> The aggregate identity type is only used by the command serviceand for calculating the [stream name](https://eventuous.dev/docs/persistence/aggregate-stream) for loading and saving events.
I will be unaware of this generated Id and thus unable to call for example card charging.
I could generate `CardId` in command service's code, but hen again I won't be able to load/store aggregate.

What should I do? I don't want the `CardId` to be generated by client itself. Maybe `AggregateId` should be accessible through aggregate / aggregate state? What am I missing?

**Comment by alexeyzimarev** [comment]
Generating ids _when_ handling commands is not the best idea. Why? Because it makes the code not idempotent. If for any reason the command is retried, you get two distinct ids for the same thing. Definitely not something you want.

If you _really_ want to generate identity in the command handler, you can add it to state and then look at the value returned by `Handle`. It's a function.
